File: C++/DFCI-Lib/Carre100Km.cpp
//
//  Carre100Km.cpp
//  DFCI-Lib-CPP
//
//  Created by Patrice Trognon on 28/10/2015.
//  Copyright © 2015 Patrice Trognon. All rights reserved.
//

#include "Carre100Km.hpp"

#include "Globals.h"

#include "Coord.hpp"
#include "Carre20Km.hpp"

#include "LambertII.hpp"
#include "LatLng.hpp"

#include "DFCICacheManager.hpp"

#include <iostream>
using namespace std;

Carre100Km::Carre100Km(string dfci) {
    nomDFCIComplet = dfci ;
    nomDFCICarre100Km = dfci.substr(0,2);
    
    string dfci100Km = nomDFCICarre100Km + "00A0.4";
    
    //---
    //   calculer le point SO en lambert
    LambertII lambert;
    LatLng latLng;
    
    lambertIIFromDFCI(dfci100Km, & lambert);
    latLngFromLambertII(lambert, & latLng);
    piCoordSO = new Coord(&latLng, &lambert);
    
    //---
    //   depuis le point SO calcul des autres points dans un carré de 100 000 m
    {
        LambertII lambertIISE(lambert.x + 100000,
                              lambert.y);
        LatLng latLngSE;
        latLngFromLambertII(lambertIISE, & latLngSE);
        piCoordSE = new Coord(&latLngSE, &lambertIISE);
    }
    
    {
        LambertII lambertIINE(lambert.x + 100000,
                              lambert.y + 100000);
        LatLng latLngNE;
        latLngFromLambertII(lambertIINE, & latLngNE);
        piCoordNE = new Coord(&latLngNE, &lambertIINE);
    }

    {
        LambertII lambertIINO(lambert.x ,
                              lambert.y + 100000);
        LatLng latLngNO;
        latLngFromLambertII(lambertIINO, & latLngNO);
        piCoordNO = new Coord(&latLngNO, &lambertIINO);
    }

    {
        LambertII lambertIICentre(lambert.x + 50000,
                                  lambert.y + 50000);
        LatLng latLngCentre;
        latLngFromLambertII(lambertIICentre, & latLngCentre);
        piCoordCentre = new Coord(&latLngCentre, &lambertIICentre);
    }
    
    for(int x=0; x<5; x++) {
        for(int y=0; y<5;y++) {
            matriceCarre20Km[x][y] = nullptr ;
        }
    }
}

Carre100Km::~Carre100Km() {
    safeDeletePtr(piCoordSO)
    safeDeletePtr(piCoordSE)
    safeDeletePtr(piCoordNE)
    safeDeletePtr(piCoordNO)
    safeDeletePtr(piCoordCentre)
    
    if ( matriceCarre20Km[0][0] == nullptr ) {
        for(int x=0; x<5; x++) {
            for(int y=0; y<5;y++) {
                safeDeletePtr(matriceCarre20Km[x][y])
            }
        }
    }
}

void
Carre100Km::performDetails() {
    if ( matriceCarre20Km[0][0] == nullptr ) {
        char LambertX[] = "02468";
        char LambertY[] = "02468";
        for(int x=0; x<5; x++) {
            for(int y=0; y<5;y++) {
                char dfci[9];
                sprintf(dfci, "%s%c%cA0.4",nomDFCICarre100Km.c_str(),LambertX[x],LambertY[y]);
                matriceCarre20Km[x][y] = new Carre20Km(dfci,this);
            }
        }
        
        for(int x=0; x<5; x++) {
            for(int y=0; y<5;y++) {
                matriceCarre20Km[x][y]->performDetails();
            }
        }
    }
}

Carre20Km*
Carre100Km::carre20Km(int x, int y) {
    return matriceCarre20Km[x][y] ;
}

Carre100Km*
Carre100Km::northCarre100Km() {
    string nameNorth = northDFCIName();
    Carre100Km *carre100Km = DFCICacheManager::getInstance()->carre100KmFromDFCI(nameNorth);
    if ( carre100Km == nullptr ) {
        carre100Km = new Carre100Km(nameNorth);
    }
    return carre100Km ;
}

Carre100Km*
Carre100Km::northEastCarre100Km() {
    return nullptr ;
}

Carre100Km*
Carre100Km::eastCarre100Km() {
    return nullptr ;
}

Carre100Km*
Carre100Km::southEastCarre100Km() {
    return nullptr ;
}

Carre100Km*
Carre100Km::southCarre100Km() {
    return nullptr ;
}

Carre100Km*
Carre100Km::southWestCarre100Km() {
    return nullptr ;
}

Carre100Km*
Carre100Km::westCarre100Km() {
    return nullptr ;
}

Carre100Km*
Carre100Km::northWestCarre100Km() {
    return nullptr ;
}

void
Carre100Km::dumpWithIndentLevel(int indentLevel) {
    cout << indentWithIndentLevel(indentLevel) << "Carre100Km {" << endl ;
    cout << indentWithIndentLevel(indentLevel) << "    dfci : " << nomDFCICarre100Km << endl ;
    dumpCoordWithPrompt(*piCoordCentre, "coordCentre", indentLevel+2);
    dumpCoordWithPrompt(*piCoordSO, "coordSO", indentLevel+2);
    dumpCoordWithPrompt(*piCoordNO, "coordNO", indentLevel+2);
    dumpCoordWithPrompt(*piCoordNE, "coordNE", indentLevel+2);
    dumpCoordWithPrompt(*piCoordSE, "coordSE", indentLevel+2);
    if ( matriceCarre20Km[0][0] != nullptr ) {
        for(int x=0; x<5; x++) {
            for(int y=0; y<5;y++) {
                matriceCarre20Km[x][y]->dumpWithIndentLevel(indentLevel+1);
            }
        }
    }
    cout << indentWithIndentLevel(indentLevel) << "}" << endl ;
}

File: Java/com/ptro/dfci/Carre100Km.java
package com.ptro.dfci;

import java.util.ArrayList;
import static com.ptro.dfci.SubString.subString;

public class Carre100Km extends CarreDFCI {
	private		String		nomDFCIComplet;
	private		String		nomDFCICarre100Km;
	private		Coord		coordCentre;
	private		Coord		coordSO;
	private		Coord		coordNO;
	private		Coord		coordNE;
	private		Coord		coordSE;
	
	private		ArrayList<Carre20Km>	matriceCarres20Km;
	
	public Carre100Km(String dfci) {
		nomDFCIComplet = dfci ;
		nomDFCICarre100Km = subString(dfci, 0, 2);
		
		String dfci100Km = nomDFCICarre100Km + "00A0.4";
		
		LambertII lambert = lambertIIFromDFCI(dfci100Km);
		LatLng latLng = latLngFromLambertII(lambert);
		coordSO = new Coord(latLng, lambert);
		
		{
			LambertII lambertIISE = new LambertII(lambert.getX() + 100000, 
												  lambert.getY());
			LatLng latLngSE = latLngFromLambertII(lambertIISE);
			coordSE = new Coord(latLngSE, lambertIISE);
		}

		{
			LambertII lambertIINE = new LambertII(lambert.getX() + 100000, 
												  lambert.getY() + 100000);
			LatLng latLngNE = latLngFromLambertII(lambertIINE);
			coordNE = new Coord(latLngNE, lambertIINE);
		}

		{
			LambertII lambertIINO = new LambertII(lambert.getX() , 
												  lambert.getY() + 100000);
			LatLng latLngNO = latLngFromLambertII(lambertIINO);
			coordNO = new Coord(latLngNO, lambertIINO);
		}

		{
			LambertII lambertIICentre = new LambertII(lambert.getX() + 50000, 
												  	  lambert.getY() + 50000);
			LatLng latLngCentre = latLngFromLambertII(lambertIICentre);
			coordCentre = new Coord(latLngCentre, lambertIICentre);
		}

		matriceCarres20Km = null ;
	}
	
	public void performDetails() {
		Carre20Km carre20Km ;
		
		if ( matriceCarres20Km == null ) {
			matriceCarres20Km = new ArrayList<Carre20Km>();
			String LambertX = "02468";
			String LambertY = "02468";
			
	        for(int x=0; x<5; x++) {
	            for(int y=0; y<5;y++) {
	            	String dfci = nomDFCICarre100Km + LambertX.charAt(x) + LambertY.charAt(y);
	                carre20Km = new Carre20Km(dfci,this);
	                matriceCarres20Km.add(carre20Km);
	            }
	        }

	        for(Carre20Km c20Km : matriceCarres20Km) {
	        	c20Km.performDetails();
	        }
		}
	}
	
	public Carre20Km carre20Km(int x, int y) {
		Carre20Km carre20Km ;
		
		performDetails();
		
		carre20Km = matriceCarres20Km.get(x*5+y);
		
		return carre20Km;
	}
	
	public void dumpWithIndentLevel(int indentLevel) {
		System.out.printf("%sCarre100Km {\n",indentWithIndentLevel(indentLevel));
		System.out.printf("%s    dfci : %s\n",indentWithIndentLevel(indentLevel),nomDFCICarre100Km);
		dumpCoordWithPrompt(coordCentre, "coordCentre", indentLevel+2);
		dumpCoordWithPrompt(coordSO, "coordSO", indentLevel+2);
		dumpCoordWithPrompt(coordNO, "coordNO", indentLevel+2);
		dumpCoordWithPrompt(coordNE, "coordNE", indentLevel+2);
		dumpCoordWithPrompt(coordSE, "coordSE", indentLevel+2);
		if ( matriceCarres20Km != null ) {
	        for(Carre20Km c20Km : matriceCarres20Km) {
	        	c20Km.dumpWithIndentLevel(indentLevel+1);
	        }
		}
		System.out.printf("%s}\n",indentWithIndentLevel(indentLevel));
	}
	
}
